### PR TITLE
refactor: use high watermark to finish backfill faster

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
step 2 & 3 of https://github.com/risingwavelabs/risingwave/issues/18299, see there for the motivation and a state diagram of the new design.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by xxchan** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18342?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#18342** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18342?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#18297** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18297?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#18296** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18296?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @xxchan and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: e2e_test/source_inline/kafka/shared_source.slt [hunk 1]
```diff
@@ -41,21 +64,24 @@ create materialized view mv_1 as select * from s0;
...
-# SourceExecutor's ingestion started, but it only starts from latest.
+# SourceExecutor's ingestion started, but it only starts from latest (offset 1).
 system ok
 internal_table.mjs --name s0 --type source
 ----
 (empty)
 
 
-# offset 0 must be backfilled, not from upstream.
+# SourceBackfill starts from offset 0, with backfill_info: HasDataToBackfill { latest_offset: "0" } (decided by kafka high watermark).
```
Comment (chenzl25) [inline e2e_test/source_inline/kafka/shared_source.slt:74]: In this case, what will the high watermark value be?

**Review by chenzl25 — APPROVED** [review]
LGTM! Great work

File: e2e_test/source_inline/kafka/shared_source.slt [hunk 2]
```diff
@@ -21,7 +44,7 @@ create source s0 (v1 int, v2 varchar) with (
   scan.startup.mode = 'earliest'
 ) FORMAT PLAIN ENCODE JSON;
 
-query I
+query ?
```
[the remark this answers is not in the document]
Comment (xxchan) [reply e2e_test/source_inline/kafka/shared_source.slt:47]: currently `sqllogictest --override` will produce `?`. Actually the character doesn't have any meaning now, any character will pass test..

File: e2e_test/source_inline/kafka/shared_source.slt (hunk 1 above)
Comment (xxchan) [reply e2e_test/source_inline/kafka/shared_source.slt:74]: 1

**Pull request merged**